### ISSUE TITLE
loss regularization

### ISSUE THREAD

**Issue by linhanxiao** [facebookresearch/fairseq-lua#107]
I want to add regularization item to the entropyloss，should I directly add  lamda×penal to the entropyloss in torch?，or should I create new Criterion Class inherited from nn.Criterion? My
regularization is not L2 regularization @jgehring @ynd @michaelauli

**Comment by michaelauli** [comment]
If you add something to CrossEntropyCriterion, then you may want to inherit from it, rather than from nn.Criterion.

**Comment by linhanxiao** [comment]
thank you for your answer @michaelauli ,i would try it

**Comment by linhanxiao** [comment]
in torch，it need to inherit from nn.CrossEntropyCriterion to impletment model regularization for classfication problem,isn't it? @ @michaelauli

**Comment by michaelauli** [comment]
Yes, I would start like this.

**Comment by linhanxiao** [comment]
my regularization equation is totalloss=entropyloss+lamda∗‖AA^T −I‖^2,the A is one network layer output, so in my new class inherited from CrossEntropyCriterion,I only need to change updateOutput method by adding  lamda*||AA^T-l||^2  to self.output   ,is that right?

**Comment by linhanxiao** [comment]
@michaelauli  i don't need to change updateGradOutput method ,is that right?

**Comment by linhanxiao** [comment]
@michaelauli ,I don't need to change updateGradInput method ,is it right?

**Comment by linhanxiao** [comment]
@michaelauli  if i only change updateOuput method，how does the regulariztion information in totalloss backward

**Comment by linhanxiao** [comment]
@michaelauli @jgehring 
the following is my loss regularization code,I directly added the penal item to crit.output,is it right for model regularization?
net:forward(sample.input)
crit:forward(net.output, sample.target)
local A
for _,b in ipairs(_G.model.selfattentivesoftmax) do
    A=b.output
end
local B=A:clone()
for i=1,B:size(1) do
     A=B[i]:clone()
    local AAT=torch.mm(A,A:t())
    local I=torch.eye(A:size(1))
    local P=torch.norm( AAT - I, 2 )
    local penal=P*P
    penal = penal/A:size(2)
     crit.output=crit.output+_G.model.selfattentivelamda*penal
end
crit:backward(net.output, sample.target)
net:backward(sample.input, crit.gradInput)

**Comment by linhanxiao** [comment]
@michaelauli @jgehring 
 the following is my loss regularization code,I directly added the penal item to crit.output,is it right for model regularization?
net:forward(sample.input)
crit:forward(net.output, sample.target)
local A
for _,b in ipairs(_G.model.selfattentivesoftmax) do
A=b.output
end
local B=A:clone()
for i=1,B:size(1) do
A=B[i]:clone()
local AAT=torch.mm(A,A:t())
local I=torch.eye(A:size(1))
local P=torch.norm( AAT - I, 2 )
local penal=P*P
penal = penal/A:size(2)
crit.output=crit.output+_G.model.selfattentivelamda*penal
end
crit:backward(net.output, sample.target)
net:backward(sample.input, crit.gradInput)

**Comment by michaelauli** [comment]
@linhanxiao, your regularization does not change parameters, e.g., see https://github.com/torch/demos/blob/master/train-a-digit-classifier/train-on-mnist.lua#L220-L225
Please check in one of the torch forums as your question is general.
